File: CMGTools/RootTools/src/Plot.cc
#include "CMGTools/RootTools/interface/Plot.h"
#include "CMGTools/RootTools/interface/TmpHistName.h"
#include "CMGTools/RootTools/interface/Styles.h"

#include <TH1F.h> 
#include <THStack.h> 
#include <TLegend.h> 

using namespace std;

Styles allStyles;

std::vector<const Style*> Plot::styles_;

// bool Plot::writeIdInfo_ = true;


Plot::Plot() : id_(""), styleIndex_(0), stack_(0), legend_(0), lxmin_(0.5), lymin_(0.75), lxmax_(1), lymax_(1) {}

Plot::Plot( const char* id ) : id_(id), styleIndex_(0), stack_(0), legend_(0), lxmin_(0.5), lymin_(0.75), lxmax_(1), lymax_(1)  {}

void Plot::AddStyle( const Style* style) {
  styles_.push_back( style );
}

void Plot::ClearStyles() {
  styles_.clear();
}

void Plot::AddHistogram(TH1* histo, const StyleAndOpt& styleAndOpt ) {

  cout<<"Plot: AddHistogram : name="<<histo->GetName()<<" , title="<<histo->GetTitle()<<endl;
  
  string name = StripOffExtension( string( histo->GetName() )  );
  histograms_[ name ] = make_pair< TH1*, StyleAndOpt >(histo, styleAndOpt ) ;
}


void Plot::SetStyle(const char* histName, const StyleAndOpt& styleAndOpt ) {

  HI hi = histograms_.find( histName );
  if( hi!=histograms_.end() ) {
    hi->second.second = styleAndOpt;
  }
  else {
    cout<<"Plot::SetStyle : cannot find histogram "<<histName<<endl;
    cout<<"available histograms are: "<<endl;
    Print( cout );
  }
}




void Plot::Draw( bool stack ) {
  
  if(stack_) delete stack_;
  stack_ = new THStack("stack", "");
  
  if(legend_) delete legend_;
  legend_ = new TLegend(lxmin_, lymin_, lxmax_, lymax_ );

  vector< HI > unStackedHistos; 

  //TODO leave freedom to define the style from outside, for all plots

  std::string stackTitle; 
  std::string stackXTitle; 
  std::string stackYTitle; 
  
  TH1* histInStack = 0;
  
  for( HI hi = histograms_.begin(); hi!=histograms_.end(); ++hi) {


    TH1* histo = hi->second.first;
    StyleAndOpt styleAndOpt = hi->second.second;

    cout<<histo->GetName()<<" "<<styleAndOpt.GetAddToStack()<<endl;

    Styles::FormatHisto(histo, styleAndOpt.GetStyle() );
    
    // in fact, should create the legend at the very end. 
    legend_->AddEntry( histo, 
		       hi->first.c_str() );

    if( !styleAndOpt.GetAddToStack() ) {
      // will be drawn at the end
      unStackedHistos.push_back(hi);
    }
    else {
      cout<<"adding to stack histo: name="<<histo->GetName()
	  <<" , title="<<histo->GetTitle()
	  <<" , xtitle="<<histo->GetXaxis()->GetTitle()<<endl;
      stack_->Add( histo );
      
      histInStack = histo;
      
      stackTitle = histo->GetTitle();
      stackXTitle = histo->GetXaxis()->GetTitle();
      stackYTitle = histo->GetYaxis()->GetTitle();
      
      //       stack_->SetTitle( histo->GetTitle() );
      //       cout<<"stack axis "<<stack_->GetXaxis()<<endl;
      //       stack_->GetXaxis()->SetTitle( histo->GetXaxis()->GetTitle() );
    }
  }

  // drawing the first unstacked histogram as a support histo
  // if the stack is drawn first, a graphics bug makes the results 
  // ugly (problem with the axes)

  string opt = "";
  if( !stack ) opt += "NOSTACK";
  
  if( !unStackedHistos.empty() ) {
//     if( writeIdInfo_ )
//       unStackedHistos[0]->second.first->SetTitle(id_.c_str() );
    unStackedHistos[0]->second.first->Draw( unStackedHistos[0]->second.second.GetOpt().c_str() );
    opt += ",A,HIST,SAME";
  }
  else {
    histInStack->Draw("NOAXIS");
    opt += ",A,HIST,SAME";    
  }

//   else{
//     cout<<"stack title: "<<stack_->GetTitle()<<endl;
//     if( writeIdInfo_ && std::string(stack_->GetTitle()).empty() )
//       stack_->SetTitle(id_.c_str() );
//   } 

  stack_->Draw(opt.c_str() );
  stack_->SetTitle( stackTitle.c_str() );
  stack_->GetXaxis()->SetTitle( stackXTitle.c_str() );
  stack_->GetYaxis()->SetTitle( stackYTitle.c_str() );
  

  // superimposing unstacked histograms

  for( unsigned id=0; id<unStackedHistos.size(); ++id ) {
    
    HI& unStacked = unStackedHistos[id];
    
//     const Style* style = NextStyle(); 
//     if(style)
//       Styles::FormatHisto(unStacked->second.first, style);    
//     legend_->AddEntry( unStacked->second.first, 
// 		       StripOffExtension( unStacked->first ).c_str() );
    unStacked->second.first->Sumw2();
    string usOpt = unStacked->second.second.GetOpt();
    usOpt += ",same";
    unStacked->second.first->Draw(usOpt.c_str());
  }
  
  // superimposing the legend
  
  legend_->Draw("same");
  
//   if(writeIdInfo_) {
//     //     idInfo_.SetNDC( true );
//     idInfo_.DrawLatex(0.02,0.95, id_.c_str() );
//   }
}


bool Plot::IsData(const string& histName) const {
  const std::string::size_type pos = histName.find("DATA"); 
  if(pos!=string::npos ) 
    return true;
  else return false;
}

std::string Plot::StripOffExtension( const string& histName) const {
  const std::string::size_type pos = histName.find("_"); 
  if(pos!=string::npos ) 
    return histName.substr(0, pos) ;
  else return histName;
  
}

void Plot::Print(std::ostream& out) const {
  for( CHI hi = histograms_.begin(); hi!=histograms_.end(); ++hi) {
    out<<hi->first<<endl;
  }
}


TH1* Plot::GetHistogram(const char* histName) {
  HI hi = histograms_.find( histName );
  if( hi!=histograms_.end() ) 
    return hi->second.first;
  else 
    return 0;
}

const Style* Plot::NextStyle() {
  if( styleIndex_ >= styles_.size() ) return 0;
  else {
    const Style* style = styles_[styleIndex_];
    styleIndex_++;
    return style;
  }
}

File: CMGTools/RootTools/python/Plot.py
from CMGTools.RootTools.Legend import *
from ROOT import TH1, THStack

from operator import itemgetter, attrgetter
import copy

class Histogram:
    def __init__(self, name, obj, layer=0, stack=True):
        self.name = name 
        self.obj = obj
        self.layer = layer
        self.stack = stack
    
    def __str__(self):
        return self.name + ' / ' + self.obj.GetName() + ', L=' + str(self.layer)

    def AddEntry(self, legend, legendLine=None):
        if legendLine == None:
            legendLine = self.name
        legend.AddEntry(self.obj, legendLine)

    def Draw(self):
        self.obj.Draw()

    def DrawNormalized(self):
        self.obj.DrawNormalized()

    def Normalize(self):
        self.obj.Scale( 1/self.obj.Integral() )


class Stack:
    def __init__(self, name):
        self.name = name
        self.hists = []
        self.integral = 0
        
    def Add(self, hist):
        self.hists.append( copy.deepcopy(hist) )
        self.integral += hist.Integral()
        
    def Draw(self, opt=''):
        if len( self.hists )==0:
            return
        self.obj = THStack(self.name,'')
        for hist in self.hists:
            self.obj.Add(hist)
        self.obj.Draw( opt )
        self.updateTitles()

    def DrawNormalized(self, opt ):
        if len( self.hists )==0:
            return
        integral = 0
        self.normHists = []
        self.obj = THStack(self.name,'')
        for hist in self.hists:
            normHist = copy.deepcopy(hist)
            normHist.Scale( 1/self.integral )
            self.normHists.append( normHist )
            self.obj.Add( normHist)
        self.obj.Draw( opt )
        self.updateTitles()

    def Normalize(self):
        for hist in self.hists:
            hist.Scale( 1/self.integral )        

    def Divide(self, otherHist):
        for hist in self.hists:
            hist.Divide(otherHist)
    
    def updateTitles( self ):
        if len( self.hists )==0:
            return        
        self.obj.GetXaxis().SetTitle( self.hists[0].GetXaxis().GetTitle() )
        self.obj.GetYaxis().SetTitle( self.hists[0].GetYaxis().GetTitle() )
        

class Plot:
    def __init__(self, name):
        self.histosDict = {}
        self.histos = []
        self.name = name
        self.legend = TLegend(0.5,0.7,1,1)
        self.legendOn = True
        self.stack = None
        
    def addHistogram(self, name, histo, layer=0, legendLine=None):
        tmp = Histogram(name, histo, layer)
        self.histos.append( tmp )
        self.histosDict[name] = tmp
        tmp.AddEntry( self.legend, legendLine)
        
    def sortedHistograms(self):
        byLayer = sorted( self.histos, key=attrgetter('layer') )
        return byLayer

    def histogram(self, name):
        return self.histosDict[name]

    def normalize(self):
        for hist in self.histos:
            hist.obj.Scale( 1/hist.obj.Integral() )

##     def normalizeStack(self):
##         if self.stack == None:
##             return
##         for hist in self.nostack:
##             hist.obj.Scale( 1/hist.obj.Integral() )
##         self.stackIntegral = stack.integral

    def hist(self, histName):
        return self.histosDict[histName]

    def DrawNormalized(self, opt=''):
        same = ''
        for hist in self.sortedHistograms():
            hist.obj.DrawNormalized( same + opt)
            if same == '':
                same = 'same'
        self.DrawLegend()

    def Draw(self, opt = ''):
        same = ''
        for hist in self.sortedHistograms():
            hist.obj.Draw( same + opt)
            if same == '':
                same = 'same'
        self.DrawLegend()

    def DrawLegend(self):
        if self.legendOn:
            self.legend.Draw('same')
                
    def DrawRatio(self):
        same = ''
        denom = None
        self.ratios = []
        for hist in self.sortedHistograms():
            if denom == None:
                denom = hist
                continue
            ratio = copy.deepcopy( hist )
            ratio.obj.Divide( denom.obj )
            ratio.obj.Draw( same )
            self.ratios.append( ratio )
            if same == '':
                same = 'same'
        self.DrawLegend()

    def DrawRatioStack(self,opt=''):
        denom = None
        histForRatios = []
        for hist in self.sortedHistograms():
            if denom == None:
                denom = hist
                continue
            histForRatios.append( hist )
        self.buildStack( histForRatios )
        self.stack.Divide( denom.obj )
        self.stack.Draw(opt)
        self.ratios = []
        for hist in self.nostack:
            print 'nostack ', hist
            ratio = copy.deepcopy( hist )
            ratio.obj.Divide( denom.obj )
            ratio.obj.Draw('same')
            self.ratios.append( ratio )
        self.DrawLegend()
                
    def DrawNormalizedRatioStack(self,opt=''):
        denom = None
        histForRatios = []
        for hist in self.sortedHistograms():
            if denom == None:
                denom = copy.deepcopy(hist)
                continue
            histForRatios.append( hist )
        self.buildStack( histForRatios )
        self.stack.Normalize()
        denom.Normalize()
        self.stack.Divide( denom.obj )
        self.stack.Draw(opt)
        self.ratios = []
        for hist in self.nostack:
            print 'nostack ', hist
            ratio = copy.deepcopy( hist )
            ratio.Normalize()
            ratio.obj.Divide( denom.obj )
            ratio.obj.Draw('same')
            self.ratios.append( ratio )        
        self.DrawLegend()
                
    def DrawStack(self, opt=''):
        self.buildStack(self.sortedHistograms())
        self.stack.Draw( opt )
        for hist in self.nostack:
            hist.obj.Draw('same')
        self.DrawLegend()

    def DrawNormalizedStack(self, opt=''):
        self.buildStack(self.sortedHistograms())
        self.stack.DrawNormalized( opt )
        for hist in self.nostack:
            hist.obj.DrawNormalized('same')
        self.DrawLegend()

    def buildStack(self, hists):
        self.stack = None
        self.stack = Stack(self.name+'_stack')
        self.nostack = []
        for hist in hists:
            if hist.stack:
                self.stack.Add( hist.obj )
            else:
                self.nostack.append(hist)
    
    def __str__(self):
        tmp =  'Plot: ' + self.name + '\n' 
        for hist in self.sortedHistograms():
            tmp += '\t' + str(hist) + '\n'
        return tmp

